File: 0020-ValidParentheses/Solution.java
import java.util.*;

public class Solution {
    
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (stack.isEmpty() || !validMatch(stack.peek(), c)) {
                stack.push(c);
            } else {
                stack.pop();
            }
        }
        return stack.isEmpty();
    }
    
    private boolean validMatch(char l, char r) {
        return l == '(' && r == ')'
            || l == '[' && r == ']'
            || l == '{' && r == '}';
    }

    public static void main(String[] args) {
        System.out.println();
    }

}

File: 0020-ValidParentheses/solution.py
class Solution:

    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            if c == '(' or c == '{' or c == '[':
                stack.append(c)
            elif stack:
                if c == ')':
                    if stack[-1] == '(':
                        stack.pop()
                    else:
                        return False
                elif c == '}':
                    if stack[-1] == '{':
                        stack.pop()
                    else:
                        return False
                else:  # c == ']'
                    if stack[-1] == '[':
                        stack.pop()
                    else:
                        return False
            else:
                return False
        return not stack


if __name__ == "__main__":
    s = Solution()

